akka-typed/src/main/scala/akka/typed/persistence/internal/PersistentActorImpl.scala

```diff
@@ -117,5 +103,27 @@ import akka.typed.internal.adapter.ActorRefAdapter
 
   }
 
+  private def applyEffects(msg: Any, effect: PersistentEffect[E, S], callback: Option[Unit ⇒ Unit] = None): Unit = effect match {
+    case CompositeEffect(mainEffect, chainedEffects) ⇒
+      applyEffects(msg, mainEffect, Some(_ ⇒ chainedEffects.foreach(applyChainableEffect)))
+    case Persist(event) ⇒
+      // apply the event before persist so that validation exception is handled before persisting
+      // the invalid event, in case such validation is implemented in the event handler.
+      state = applyEvent(state, event)
+      persist(event) { _ ⇒
+        callback.foreach(_.apply(()))
```

Comment:
this `callback` isn't used now, I think

Reply:
It's filled in the recursive call when handling the `CompositeEffect`